torchdata/stateful_dataloader/ibm_rescalable.py

```diff
@@ -0,0 +1,614 @@
+        out = {}
+        for k, v in state_dict.items():
+            v = torch.tensor(v)
+            if len(v.shape) == 0:
+                k = k + ".scalar"
+                v = v.unsqueeze(0)
+            out[k] = v
+
+        return out
+
+    def load_state_dict(self, state_dict):
+        self.setup()
+
+        # Convert back to lists and scalars
+        def detorchify(k, v):
+            v = v.tolist()
+            if ".scalar" in k:
+                k = k[:-7]
+                v = v[0]
+            return k, v
+
+        plain_dict = {}
+        for k, v in state_dict.items():
+            k, v = detorchify(k, v)
+            plain_dict[k] = v
+        state_dict = plain_dict
+
+        # Assemble logical shard states
+        # TODO: how is this handling non-resharding state_params when resharding???
+        _StatefulDataset.load_state_dict(self, state_dict)
+        # Remove all non-resharding state
+        [state_dict.pop(self.statename(n)) for n in self.state_params]
+        # Flip dict[list[any]] to list[dict[any]]
+        logical_shard_states = [{k: v[i] for k, v in state_dict.items()} for i in range(self.n_logicals)]
+
+        # Load values
+        for i in range(self.n_logicals):
+            self.data[i].load_state_dict(logical_shard_states[i])
+
+
+#### -------------------------    CHECKPOINT FUNCTIONS    ------------------------- ####
+
+
+def __pop_dstate(state, device_mesh, placements):
+    """
+    Removes worker states from the StatefulDataLoader state dict, and assembles them
+    into a separate list of dicts for distributed checkpointing.
+    """
+    dstate = state["_snapshot"]["_worker_snapshots"]
+    dstate = [dstate[f"worker_{i}"].pop("dataset_state") for i in range(len(dstate))]
+    # Flip list[dict[tensor]] to dict[list[tensor]], and concat
+    dstate = {k: torch.cat([d[k] for d in dstate], 0) for k in dstate[0]}
+    # Construct dtensors from tensors
+    dstate = {
+        k: dtensor.DTensor.from_local(
+            v,
+            device_mesh,
+            placements,
+        )
+        for k, v in dstate.items()
+    }
+    return dstate
+
+
+def save_distributed_state_dict(
+    loader: StatefulDataLoader,
+    path: str,
+    device_mesh=None,
+):
+    """
+    Retrieves dataloader state dict, and separates worker states from loader state.
+    Loader state is not rescalable, and is saved using normal torch.save.
+    It is discarded when rescaling.
+    Rescalable worker states are compiled into a dtensor across ranks, and saved 
+    using pytorch distributed checkpointing.
+    """
+    rank = loader.dataset.rank
+    state = deepcopy(loader.state_dict())
+    dstate = __pop_dstate(state, device_mesh, [dtensor.placement_types.Shard(0)])
+    # Write distributed state dict
+    writer = checkpoint.FileSystemWriter(path)
+    checkpoint.save(
+        dstate,
+        writer,
+    )
+    # Write nondistributed state dict
+    torch.save(state, os.path.join(path, f"__nondist_cp_{rank}.pth"))
```

Comment:
Shall we also make this state part of the main checkpoint? We can use the torch.save serialization (output is bytestream), to store the state as part of the DCP checkpoint.

buff = io.BytesIO()
torch.save(state, buff) # or we can serialize individual keys in the state-dict. But no strong need.
buff.seek(0)

assume the unique key is 'trainer_dataloader_state_rank_k' -> "

update the dstate with new key -> value.
checkpoint.save(dstate)

Reply:
Some context on how DCP handles regular ( non-dtensor) tensor and non-tensor values found in state dict.

1\ regular tensors ( not distributed tensors) are saved against a unique fqn. The state dict can have { 'foo.bar.1' : torch.ones((2,3), dtype=float). As long as the fqns do not collide between different ranks we are good.

2\ DCP can save/load non-tensor values found in state dict as well.  That means the value can be, a string, list of ints or a ByteIO object. Under the hood DCP uses torch.save/torch.load to serialize/deserialize these non-tensor (we call them blobs) values during checkpoint save/load. E.g. UT -> https://github.com/pytorch/pytorch/blob/11bb94b7eaf272da8e1f1dfd94c3d8872247e895/test/distributed/checkpoint/test_file_system_checkpoint_cpu.py#L367


Therefore, we  can use the dataloader state-dict values as it is, to save them as part of the DCP checkpoint.